pkg/resourcemanager/mysql/mysqlhelper.go

```diff
@@ -0,0 +1,164 @@
+func ConnectToSqlDB(ctx context.Context, driverName string, fullServer string, database string, port int, user string, password string) (*sql.DB, error) {
+	connString := fmt.Sprintf("%s:%s@tcp(%s:%d)/%s?tls=true&interpolateParams=true", user, password, fullServer, port, database)
+
+	db, err := sql.Open(driverName, connString)
+	if err != nil {
+		return db, err
+	}
+
+	err = db.PingContext(ctx)
+	if err != nil {
+		return db, fmt.Errorf("error ping the mysql db:  %v", err)
+	}
+
+	return db, err
+}
+
+// ConnectToSQLDBAsCurrentUser connects to the SQL DB using the specified MSI ClientID
+func ConnectToSQLDBAsCurrentUser(
+	ctx context.Context,
+	driverName string,
+	fullServer string,
+	database string,
+	port int,
+	user string,
+	clientID string) (*sql.DB, error) {
+
+	tokenProvider, err := iam.GetMSITokenProviderForResourceByClientID("https://ossrdbms-aad.database.windows.net", clientID)
+	if err != nil {
+		return nil, err
+	}
+
+	// In our case we can't pass the provider directly so we just invoke it and get the token and use that
+	token, err := tokenProvider()
+	if err != nil {
+		return nil, err
+	}
+
+	// See https://docs.microsoft.com/en-us/azure/mysql/howto-connect-with-managed-identity
+	// As noted here https://docs.microsoft.com/en-us/azure/mysql/howto-configure-sign-in-azure-ad-authentication#compatibility-with-application-drivers
+	// we must specify allowCleartextPasswords to pass a token
+	connString := fmt.Sprintf("%s:%s@tcp(%s:%d)/%s?tls=true&allowCleartextPasswords=true&interpolateParams=true", user, token, fullServer, port, database)
+
+	db, err := sql.Open(driverName, connString)
+	if err != nil {
+		return db, err
+	}
+
+	err = db.PingContext(ctx)
+	if err != nil {
+		return db, fmt.Errorf("error ping the mysql db:  %v", err)
+	}
+
+	return db, err
+}
+
+func GrantUserRoles(ctx context.Context, user string, database string, roles []string, db *sql.DB) error {
+	var errorStrings []string
+	if err := helpers.FindBadChars(user); err != nil {
+		return fmt.Errorf("problem found with username: %v", err)
+	}
+
+	for _, role := range roles {
+
+		if err := helpers.FindBadChars(role); err != nil {
+			return fmt.Errorf("problem found with role: %v", err)
+		}
+
+		// Due to how go-mysql-driver performs parameter replacement, it always wraps
+		// string parameters in ''. That doesn't work for this query because some of
+		// our parameters are actually SQL keywords or identifiers (backticks). Admittedly
+		// protecting against SQL injection here is probably pointless as we're giving the caller
+		// permission to create users, which means there's nothing stopping them from creating
+		// an administrator user and then doing whatever they want without SQL injection.
+		// See https://github.com/go-sql-driver/mysql/blob/3b935426341bc5d229eafd936e4f4240da027ccd/connection.go#L198
+		// for specifics of what go-mysql-driver supports.
+		tsql := fmt.Sprintf("GRANT %s ON `%s`.* TO '%s'", role, database, user)
```

Comment:
I'd feel better if you did this in two phases - `fmt.Sprintf` in the role and db since they can't be parameters, and then pass the username as a query parameter. You're right that ultimately if someone can call this then they can make their own user but there still... Not sure really.

Reply:
Done as you mentioned. Left the comment as well... here be dragons for sure... we may want to think of a better way to do this sort of thing.